core/src/main/java/kafka/log/remote/RemoteLogManager.java

```diff
@@ -1073,13 +1088,28 @@ void cleanupExpiredRemoteLogSegments() throws RemoteStorageException, ExecutionE
                         .iterator();
                 while (epochsToClean.hasNext()) {
                     int epoch = epochsToClean.next();
+                    List<RemoteLogSegmentMetadata> listOfSegmentsToBeCleaned = new ArrayList<>();
                     Iterator<RemoteLogSegmentMetadata> segmentsToBeCleaned = remoteLogMetadataManager.listRemoteLogSegments(topicIdPartition, epoch);
                     while (segmentsToBeCleaned.hasNext()) {
                         if (isCancelled() || !isLeader()) {
                             return;
+                        } else {
+                            RemoteLogSegmentMetadata nextSegmentMetadata = segmentsToBeCleaned.next();
+                            sizeOfDeletableSegmentsBytes += nextSegmentMetadata.segmentSizeInBytes();
+                            listOfSegmentsToBeCleaned.add(nextSegmentMetadata);
                         }
+                    }
+                    segmentsLeftToDelete += listOfSegmentsToBeCleaned.size();
+                    brokerTopicMetrics.recordRemoteDeleteBytesLag(partition, sizeOfDeletableSegmentsBytes);
+                    brokerTopicMetrics.recordRemoteDeleteSegmentsLag(partition, segmentsLeftToDelete);
+                    for (RemoteLogSegmentMetadata segmentMetadata : listOfSegmentsToBeCleaned) {
                         // No need to update the log-start-offset even though the segment is deleted as these epochs/offsets are earlier to that value.
-                        remoteLogRetentionHandler.deleteLogSegmentsDueToLeaderEpochCacheTruncation(earliestEpochEntry, segmentsToBeCleaned.next());
+                        if (remoteLogRetentionHandler.deleteLogSegmentsDueToLeaderEpochCacheTruncation(earliestEpochEntry, segmentMetadata)) {
```

Comment:
Also a very crucial catch - all of this was meant to be in the else condition. The if condition would have terminated should the current replica not be the leader - I will change this. I will also put a note to add a test in the upcoming future, because nothing broke when I accidentally changed the behaviour 😬 

For my understanding, why is this code branch any more in-sync or async than the other deletions? As far as I can tell both eventually call `deleteRemoteLogSegment` - am I missing something very obvious?

Reply:
> For my understanding, why is this code branch any more in-sync or async than the other deletions? As far as I can tell both eventually call deleteRemoteLogSegment - am I missing something very obvious?

My understanding with `RemoteDeleteLagBytes` and `RemoteDeleteLagSegments` metrics is that it gives the user a gist of remote segments that are eligible for deletion but they couldn't be deleted now due to degradation in the secondary storage.

Here, we are calculating the lag for each segment and deleting the same. If the secondary storage throws RemoteStorageException, then the lag calculated becomes invalid.

Reply:
Sorry, I am afraid I do not understand why would the lag become invalid. In my head, we start from everything we need to delete and we gradually try to delete it. If at any point in time secondary storage throws a tantrum then the lag is the difference between what we were last successfully able to delete and what we should have deleted up to. On the next deletion iteration we should pick up the segments we were unable to delete and they would be included in the calculation again. Am I missing something obvious here?